src/devices/Mcp23xxx/tests/Mcp23xxxTest.cs

```diff
@@ -96,7 +98,10 @@ public I2cDeviceMock(int ports, I2cConnectionSettings settings = null)
             public override byte ReadByte() => throw new NotImplementedException();
         }
 
-        public class Mcp23xxxMock
+        /// <summary>
+        /// Mock the behavior of the chip
+        /// </summary>
+        public class Mcp23xxxChipMock
```

Comment:
Question: I have not done much mocking (but hopefully plan to as I learn more in this space).  When it comes to bindings, is it preferred to create similar to what you've done here or is it accepted to use other packages like [Moq](https://www.nuget.org/packages/moq/)?

Reply:
I discussed this point briefly with @joperezr. I think we should be using Moq for tests, I just didn't want to consume a specific library without building some consensus first.

We don't do a ton of mocking in CoreFX as we rely on the hardware. In my previous group we did. The recommendation is to use the mocking framework as it helps describe what you're actually testing and allows you to do more extensive validation easily (like only this method was called and it was only called once).

In this particular case I'm not mocking another API I'm simulating some hardware. I probably should have called it `Mcp32xxxChipSimulator` to be clearer? I could have put this functionality in each of the classes I am mocking, and in fact did to begin with.  I found it harder to track what I was doing and understand what was going on in the tests.

Reply:
We might want to come up with practices around using Moq and bindings to help make consistent.  Maybe some thoughts around the main pin types (Gpio, Pwm, I2c, Spi, etc.) would be good.  Of course, actually trying it on the hardware would tell us if it really worked or not (with timing, connection, etc.), but I could see the "simulators" being helpful... especially getting started on coding and studying the chip datasheet (while waiting on the chips to arrive in the mail :smile:)

Reply:
> but I could see the "simulators" being helpful.

They do provide quite a bit of value. Actually running on hardware is the proof in the pudding, of course. We can't, however, reasonably run automation on real devices (well, not many of them for sure).  Simulating the hardware can help validate behavior and avoid regressions. While we'll undoubtedly have bugs in our simulations, those will get fixed over time and we'll have much higher confidence in the state of our code base before it gets to the unbounded permutations of the real world.

One way I think of it sometimes is as an emulated or low fidelity record/playback system.